mcs/class/System/System.Net/EndPointListener.cs

```diff
@@ -82,28 +82,26 @@ public EndPointListener (HttpListener listener, IPAddress addr, int port, bool s
 			get { return listener; }
 		}
 
-		static void OnAccept (object sender, EventArgs e)
-		{
-			SocketAsyncEventArgs args = (SocketAsyncEventArgs) e;
-			EndPointListener epl = (EndPointListener) args.UserToken;
-			Socket accepted = null;
-			if (args.SocketError == SocketError.Success) {
-				accepted = args.AcceptSocket;
-				args.AcceptSocket = null;
-			}
+                static void Accept (Socket socket, SocketAsyncEventArgs e)
+                {
+                        e.AcceptSocket = null;
+                        var asyn = socket.AcceptAsync(e);
+                        if (!asyn) {
+                                ProcessAccept(e);
+                        }
+                }
 
-			try {
-				if (epl.sock != null)
-					epl.sock.AcceptAsync (args);
-			} catch {
-				if (accepted != null) {
-					try {
-						accepted.Close ();
-					} catch {}
-					accepted = null;
-				}
-			} 
 
+                static void ProcessAccept (SocketAsyncEventArgs args)
+                {
+                        Socket accepted = null;
+                        if (args.SocketError == SocketError.Success)
+                                accepted = args.AcceptSocket;
+
+                        EndPointListener epl = (EndPointListener) args.UserToken;
+
+
+                        Accept (epl.sock, args);
```

Comment:
What are you hoping that code to do? Your entire listener is hosed at that
point (similar to the issue this patch resolves) as you have broken the
async loop, no future accept will happen. Accepted may also still be a
valid socket. I am not clear what the benefit of setting the accepted
socket to null would be.

On Mon, Mar 7, 2016 at 2:11 PM, Ludovic Henry <notifications@github.com>
wrote:

> In mcs/class/System/System.Net/EndPointListener.cs
> <https://github.com/mono/mono/pull/2713#discussion_r55194917>:
>
> >
> > +                static void ProcessAccept (SocketAsyncEventArgs args)
> > +                {
> > +                        Socket accepted = null;
> > +                        if (args.SocketError == SocketError.Success)
> > +                                accepted = args.AcceptSocket;
> > +
> > +                        EndPointListener epl = (EndPointListener) args.UserToken;
> > +
> > +
> > +                        Accept (epl.sock, args);
>
> We are missing the try { ... } catch { ...; accepted = null; } from
> before.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mono/mono/pull/2713/files#r55194917>.
>



-- 
Studying for the Turing test